[Sourabh] iP

srj31 opened the pull request:
# DukePro (v1.0)

> Java is to JavaScript what car is to Carpet. 
> – Chris Heilmann

DukePro frees your mind of having to remember things you need to do. It has 

* a **text based** interface
* an *intuitive* design

All you need to do to get it,

1. clone the repo or get the jar file (use [this](https://www.google.com/) to see how to use jar files)
2. double click
3. start adding tasks :smiley:

And ~pay $10~ it's **FREE!!!**

Features: 

- [x] managing tasks (todos, events, deadlines)
- [ ] Reminders (coming soon)

If you know how to program in Java, can edit the files in `Duke.java` and make changes in the `main`  method as per the changes needed: 

```java

public class Duke {
     public static void main(String[] args) throws DukeException, IOException {
        Duke duke = new Duke();
        duke.runDuke();
    }
}

```

Commit:
Add Gradle support

srj31 committed:
level-1 ip greet, echo, exit

srj31 committed:
creating a list of commands

srj31 committed:
using inheritance Task

srj31 committed:
Level4

srj31 committed:
added ui testing

srj31 committed:
Adding Exceptions

srj31 committed:
Adding DukeExceptions

srj31 committed:
Deleting Tasks

srj31 committed:
fixing ui-testing

srj31 committed:
saving and loading to a file

srj31 committed:
reading date

srj31 committed:
merging branch-level-7

srj31 committed:
resolve conflict

srj31 committed:
Using oop principles

srj31 committed:
using duke package

srj31 committed:
moving everything inside duke

srj31 committed:
adding JUnit test

srj31 committed:
Building jar

srj31 committed:
adding header comments

srj31 committed:
using CodingStandard

srj31 committed:
adding the find feature

srj31 committed:
Fix Merge conflicts

srj31 committed:
resolved conflicts

willyamped commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+package duke;
+
+/**
+ *
+ */
```
I think there is a missing javadocs header here? 😄

calvintanwj replied:
Agreed, seems like it's walked off somewhere. 😄

willyamped commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    /**
+     * To get the time of the event stored in the list
```
Javadoc headers should start with words in the form `Returns ...`, `Sends ...`, `Adds ...` (not `Return` or `Returning` etc.)
This applies to your other javadoc headers too :)

willyamped commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,85 @@
+            for(Task task: Tasklist.dukeList) {
+                String commandLine = "";
+                if(task instanceof Deadline) {
+                    commandLine = "D | " + (task.getIsDone() ? "1 | " : "0 | ") + task.getDescription() + " | " + ((Deadline) task).getTime() + '\n';
+                } else if(task instanceof Event) {
+                    commandLine = "E | " + (task.getIsDone() ? "1 | " : "0 | ") + task.getDescription() + " | " + ((Event) task).getTime() + '\n';
+                } else {
+                    commandLine =  "T | " + (task.getIsDone() ? "1 | " : "0 | ") + task.getDescription() + '\n';
```
I think this chunk is not very reader-friendly 😬? You could try splitting them up into a few lines. Remember to follow the convention (e.g the number of tabs, etc)

willyamped commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,153 @@
+     * Remove a Task from the list
+     * @param taskIndex the index of the Task to be removed
+     */
+    void deleteTask(int taskIndex) {
```
I like that you create these separate methods 👍

willyamped reviewed (commented):
LGTM. Overall, the code is good except for some minor things

calvintanwj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+
+import duke.exceptions.DukeException;
+
+import java.io.*;
```
When you import your classes, I think it would be better to list them explicitly.
```suggestion
import java.io.IOException;
import ...
```
This adheres to the coding standards, which should be the same for your tP. Additionally, it gives excellent documentation value for the team. I believe for IntelliJ, if 3 or more classes are imported from the same package, IntelliJ automatically converts it into a wilcard import, this limit can be toggled or turned off within your IDE settings.

calvintanwj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        ui.displayWelcomeMessage();
+        String command = sc.next();
+        String description = sc.nextLine();
+        while(!command.equals("bye")) {
```
I believe there should be a space between your while keyword and condition.
```suggestion
        while (!command.equals("bye")) {
```

calvintanwj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        while(!command.equals("bye")) {
+            handleCommandExecution(command,description);
+            command = sc.next();
+            if(!command.equals("bye")) {
```
Same as above.

calvintanwj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        ui.displayByeMessage();
+    }
+
+    public void handleCommandExecution(String command, String description)throws DukeException {
```
Same as above. Missing some spaces. 
```suggestion
    public void handleCommandExecution(String command, String description) throws DukeException {
```

calvintanwj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+     * @param string    the input line
+     * @return          array containing the description and time info
+     */
+    public static String[] splitDescriptionAndTime(String string) {
```
Not too sure if I quite like the naming convention for the parameter, string. 
May I suggest something like "inputString".

calvintanwj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /**
+     * Extracts the description after splitting using splitDescriptionAndTime
+     * @param description_and_time  array after splitting the input line
+     * @return                      the description of the Task
```
For javadoc comments, I believe there should be a spacing between your description of the method and the params, like:
```suggestion
     * Extracts the description after splitting using splitDescriptionAndTime
     *
     * @param description_and_time  array after splitting the input line
     * @return                      the description of the Task
```

calvintanwj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,85 @@
+        Scanner scanner = new Scanner(savedFile);
+        while (scanner.hasNextLine()) {
+            String[] keywords = scanner.nextLine().split(" \\| ");
+            Task cur = null;
```
Perhaps instead of "cur", "currTask" may be better.

calvintanwj commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,22 @@
+public class DukeTest {
+
+    @Test
+    public void testWrongCommand() {
```
I think that this method naming convention is fine, an possible alternative you and your team could use: is **featureUnderTest_testScenario_expectedBehavior()**

calvintanwj reviewed (commented):
Overall, really good stuff 👍 Just need to be a bit careful of those spaces and indentations, especially for your if and while blocks. I forget about those sometimes as well, perhaps you can configure your editor (File > Settings > Editor > Code Style > Java) to follow the coding standards. This way, you just need to run the reformat code (ctrl + alt + j) before you commit your changes. There are also a few other cosmetic issues like javadoc comments and naming conventions for methods and arguments, but these are rather minor. All the files are packaged nicely, with most of the methods having descriptive comments.

srj31 committed:
reviewing comments

srj31 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

srj31 committed:
build and run using gradle

srj31 committed:
Merge branch 'branch-A-Gradle'

srj31 committed:
fixing reviews and adding tag

srj31 committed:
Implementing basic GUI and controls for Duke

srj31 committed:
Completing GUI

srj31 committed:
Merge branch 'branch-Level-10'

luffingluffy commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,27 @@
+package duke;
+
+/**
+ * duke.Event class to create Events for the duke list
```
I think that you are missing full stops for the JavaDocs; similarly for the rest of the code.

luffingluffy commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * If the task is marked done X is shown
```
Maybe it's better to write actively; "Shows X if the task is marked done"

luffingluffy commented on src/main/java/duke/Tasklist.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * Removing a task from the list of the tasks.
```
Similar to the rest, maybe it is better to write as "Removes a task.."

luffingluffy reviewed (commented):
I think can improve some minor JavaDocs formatting.

srj31 committed:
starting with fxml

srj31 committed:
Merge branch 'branch-Level-10'

srj31 committed:
adding assertions

srj31 committed:
Improve CodeQuality of the project

The executeCommand method has some common behaviors for certain command types
(e.g. getting the index) as well as can be refactored
(e.g. adding task messages).

The common behaviors cause code duplicaiton and the method is long and should
be refactored.

Extracting the common behavior into a function allows use to use polymorphism
when dealing with the various kinds of Tasks. For example, displaying add
message for the different kind of Tasks with one function.

Let's pull up the common behavior of displaying the message into a new
function displayAddMessage and also refactor the code for adding new
deadlines and events.

srj31 committed:
Merge pull request #1 from srj31/branch-A-Assertions

adding assertions

srj31 committed:
Merge branch 'master' into branch-A-CodeQuality

srj31 committed:
Merge pull request #2 from srj31/branch-A-CodeQuality

Improve CodeQuality of the project

srj31 committed:
Fix comment error

srj31 committed:
fixing the tag

srj31 committed:
Update command: edit task description and time

Only way to change task is to delete and add another.

Update command allows to edit task description and time easily
in one line.

Let's add another command in the executeCommand function to allow
for editing.

srj31 committed:
Add User Guide

srj31 committed:
Merge branch 'branch-A-UserGuide'

srj31 committed:
Set theme jekyll-theme-cayman

srj31 committed:
update build.gradle

srj31 committed:
Merge branch 'master' of https://github.com/srj31/ip

srj31 committed:
fix header commits

srj31 committed:
test TodoUpdateWithTime

srj31 committed:
refactor code for not very long method

srj31 committed:
Merge pull request #5 from srj31/branch-better-abstraction

refactor code for not very long method

srj31 committed:
fix switch case style
